File: LeetCode/iter1/java/821.shortest-distance-to-a-character.java
package org.ironman.leetcode;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Solution {

    public static int[] shortestToChar(String S, char C) {
        List<Integer> cIdxLst = new ArrayList<>();
        cIdxLst.add(Integer.MAX_VALUE);
        for (int i = 0; i < S.length(); i++) {
            if (S.charAt(i) == C) {
                cIdxLst.add(i);
            }
        }
        cIdxLst.add(Integer.MAX_VALUE);

        int[] ret = new int[S.length()];
        int cIdx = 1, preCIdx = cIdxLst.get(0), curCIdx = cIdxLst.get(cIdx);
        for (int i = 0; i < S.length(); i++) {
            if (i > curCIdx) {
                preCIdx = curCIdx;
                curCIdx = cIdxLst.get(++cIdx);
            }
            ret[i] = Math.min(Math.abs(preCIdx - i), Math.abs(curCIdx - i));
        }

        return ret;
    }

    public static void main(String[] args) {
        String S = "loveleetcode";
        char C = 'e';
        System.out.println(Arrays.toString(shortestToChar(S, C)));
    }
}

File: LeetCode/iter1/c++/821.shortest-distance-to-a-character.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <vector>
#include <sstream>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <queue>
#include <stack>

using namespace std;

class Solution {
public:
	vector<int> shortestToChar(string S, char C) {
		vector<int> CIdxMap;
		CIdxMap.push_back(INT_MAX);
		for (int i = 0; i < S.length(); i++) {
			if (S[i] == C) {
				CIdxMap.push_back(i);
			}
		}
		CIdxMap.push_back(INT_MAX);

		vector<int> ret;
		int CIdx = 1, preCIdx = CIdxMap[CIdx-1], curCIdx = CIdxMap[CIdx];
		for (int i = 0; i < S.length(); i++) {
			if (i > curCIdx) {
				preCIdx = curCIdx;
				curCIdx = CIdxMap[++CIdx];
			}
			
			ret.push_back(min(abs(preCIdx - i), abs(curCIdx - i)));
		}

		return ret;
	}
};

void printHelper(vector<int> v) {
	for (int num : v) {
		cout << num << " ";
	}
	cout << endl;
}

int main()
{
	Solution solution;
	//eg1
	string S = "loveleetcode";
	char C = 'e';
	printHelper(solution.shortestToChar(S, C));

	return 0;
}